dashboard/test/models/contact_rollups_pardot_memory_test.rb

```diff
@@ -32,52 +32,167 @@ class ContactRollupsPardotMemoryTest < ActiveSupport::TestCase
     ContactRollupsPardotMemory.create_new_pardot_prospects
-    assert_equal processed_contacts.length, ContactRollupsPardotMemory.count
-    processed_contacts.each do |contact|
-      refute_nil ContactRollupsPardotMemory.find_by(email: contact.email)
+
+    assert_equal 1, ContactRollupsPardotMemory.count
+    record = ContactRollupsPardotMemory.find_by(email: contact.email)
+    assert_equal({'db_Opt_In' => 'Yes'}, record&.data_synced)
+  end
+
+  test 'query_updated_contacts' do
+    assert_equal 0, ContactRollupsPardotMemory.count
+    assert_equal 0, ContactRollupsProcessed.count
+
+    base_time = Time.now.utc - 2.days
+    pardot_memory_records = [
+      {email: 'alpha', pardot_id: 1, data_synced_at: nil, data_synced: nil},
+      {
+        email: 'beta',
+        pardot_id: 2, pardot_id_updated_at: base_time - 2.days,
+        data_synced_at: base_time - 1.day, data_synced: {db_Opt_In: 'Yes'}
+      },
+      {
+        email: 'gamma',
+        pardot_id: 3, pardot_id_updated_at: base_time + 2.days,
+        data_synced_at: base_time + 1.day, data_synced: {db_Opt_In: 'Yes'}
+      },
+      # dummy records
+      {email: 'delta', pardot_id: nil},
+      {email: 'epsilon', pardot_id: 4},
+    ]
+    pardot_memory_records.each {|record| create :contact_rollups_pardot_memory, record}
+
+    # 3 cases that require updating contacts
+    processed_contact_records = [
+      # Case 1: data_synced_at is null
+      {email: 'alpha', data: {opt_in: false}, data_updated_at: base_time},
+      # Case 2: data_updated_at > data_synced_at
+      {email: 'beta', data: {opt_in: false}, data_updated_at: base_time},
+      # Case 3: pardot_id_updated_at > data_synced_at
+      {email: 'gamma', data: {opt_in: true}, data_updated_at: base_time},
+      # dummy records
+      {email: 'delta'},
+      {email: 'zeta'},
+    ]
+    processed_contact_records.each {|record| create :contact_rollups_processed, record}
+
+    # Execute SQL query
+    results = ActiveRecord::Base.connection.
+      exec_query(ContactRollupsPardotMemory.query_updated_contacts).map do |record|
+      record.slice('email', 'pardot_id_changed')
     end
+
+    expected_results = [
+      {'email' => 'alpha', 'pardot_id_changed' => nil},
+      {'email' => 'beta', 'pardot_id_changed' => 0},
+      {'email' => 'gamma', 'pardot_id_changed' => 1}
+    ]
+    assert_equal expected_results, results
   end
 
-  test 'save_sync_results' do
+  test 'update_pardot_prospects' do
+    email = 'test@domain.com'
+    last_sync_time = Time.now.utc - 7.days
+    create :contact_rollups_pardot_memory, email: email, data_synced: {db_Opt_In: 'No'}, data_synced_at: last_sync_time
+    create :contact_rollups_processed, email: email, data: {'opt_in' => true}
+
+    PardotV2.expects(:submit_batch_request).once.returns([])
+
+    ContactRollupsPardotMemory.update_pardot_prospects
+
+    record = ContactRollupsPardotMemory.find_by(email: email)
+    assert_equal({'db_Opt_In' => 'Yes'}, record&.data_synced)
```

Comment:
Why safe navigate here?

Reply:
You're right, it doesn't need safe navigator. Will fix.